src/elf_parser/mod.rs

```diff
@@ -0,0 +1,616 @@
+            .map(|remaining| remaining != 0)
+            .unwrap_or(true)
+        {
+            return Err(ElfParserError::InvalidAlignment);
+        }
+        let file_header = unsafe { &*ptr.cast::<Elf64Ehdr>() };
+
+        if file_header.e_ident.ei_mag != ELFMAG
+            || file_header.e_ident.ei_class != ELFCLASS64
+            || file_header.e_ident.ei_data != ELFDATA2LSB
+            || file_header.e_ident.ei_version != EV_CURRENT as u8
+            || file_header.e_version != EV_CURRENT
+            || file_header.e_ehsize != std::mem::size_of::<Elf64Ehdr>() as u16
+            || file_header.e_phentsize != std::mem::size_of::<Elf64Phdr>() as u16
+            || file_header.e_shentsize != std::mem::size_of::<Elf64Shdr>() as u16
+            || file_header.e_shstrndx >= file_header.e_shnum
+        {
+            return Err(ElfParserError::InvalidFileHeader);
+        }
+
+        let program_header_table_range = file_header.e_phoff as usize
+            ..std::mem::size_of::<Elf64Phdr>()
+                .saturating_mul(file_header.e_phnum as usize)
+                .saturating_add(file_header.e_phoff as usize);
+        check_that_there_is_no_overlap(&file_header_range, &program_header_table_range)?;
+        let program_header_table =
+            slice_from_bytes::<Elf64Phdr>(elf_bytes, program_header_table_range.clone())?;
+
+        let section_header_table_range = file_header.e_shoff as usize
+            ..std::mem::size_of::<Elf64Shdr>()
+                .saturating_mul(file_header.e_shnum as usize)
+                .saturating_add(file_header.e_shoff as usize);
+        check_that_there_is_no_overlap(&file_header_range, &section_header_table_range)?;
+        check_that_there_is_no_overlap(&program_header_table_range, &section_header_table_range)?;
+        let section_header_table =
+            slice_from_bytes::<Elf64Shdr>(elf_bytes, section_header_table_range.clone())?;
+
+        let mut prev_program_header: Option<&Elf64Phdr> = None;
+        for program_header in program_header_table {
+            if program_header.p_type != PT_LOAD {
+                continue;
+            }
+
+            if let Some(prev_program_header) = prev_program_header {
+                // program headers must be ascending
+                if program_header.p_vaddr < prev_program_header.p_vaddr {
+                    return Err(ElfParserError::InvalidProgramHeader);
+                }
+            }
+
+            if program_header
+                .p_offset
+                .saturating_add(program_header.p_filesz) as usize
+                > elf_bytes.len()
+            {
+                return Err(ElfParserError::OutOfBounds);
+            }
+
+            prev_program_header = Some(program_header)
+        }
+
+        let mut offset = 0usize;
+        for section_header in section_header_table.iter() {
+            if section_header.sh_type == SHT_NOBITS {
+                continue;
```

Comment:
Think we can restrict this further?

Reply:
Actually, it is the other way around. The continue statement here excludes sections which have no data in the file (SHT_NOBITS) from the restrictive checks below.

Reply:
I'm "+0" on doing those checks - but keep in mind that it's checking that the sections don't overlap with the file header, program headers and the section table. Those are all things not allowed in ELF, and I can't think of anything that would generate such files (barring corruption or intentionally malicious inputs)